core/commonTest/src/kotlinx/serialization/encoding/ReorderingCompositeEncoderTest.kt

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2017-2024 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization.encoding
+
+import kotlinx.serialization.*
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.modules.*
+import kotlin.jvm.*
+import kotlin.test.*
+
+class ReorderingCompositeEncoderTest {
+    @Test
+    fun shouldReorderWell() {
+        val first = 42
+        val second = "string"
+        val third = true
+        val value = OuterToBeReordered(Middle(Inner(second, third)), first)
+
+        val encoder = ReorderingEncoder()
+
+        OuterToBeReordered.serializer().serialize(encoder, value)
+
+        assertContentEquals(actual = encoder.encodedValues, expected = listOf(first, second, third))
```

Comment:
The test passed, no exception at all and it's encoded in the good order.

Btw, this test won't remains, but I cannot code as I have a big issue on IntelliJ, even with cache invalidation, I have everything red. This is the only project with that issue, but also the only multiplatform project that I have... Still searching, kotlin standard library completion doesn't work (mutableListOf, apply, UnsupportedOperationException), it's a nightmare 😅